aten/src/ATen/native/TensorIndexing.h

```diff
@@ -0,0 +1,110 @@
+#pragma once
+
+#include <c10/util/Optional.h>
+#include <ATen/core/TensorBody.h>
+
+namespace at {
+namespace indexing {
+
+const int64_t INDEX_MAX = std::numeric_limits<int64_t>::max();
+const int64_t INDEX_MIN = std::numeric_limits<int64_t>::min();
+
+enum class TensorIndexType { None, Ellipsis, Integer, Boolean, Slice, Tensor };
+
+constexpr c10::nullopt_t None{c10::nullopt_t::init()};
+
+struct CAFFE2_API EllipsisIndexType { EllipsisIndexType() {} };
+CAFFE2_API extern const EllipsisIndexType Ellipsis;
+
+struct CAFFE2_API Slice {
+ public:
+  Slice();
+  Slice(int64_t start, int64_t stop, int64_t step);
+
+  const int64_t& start() const;
+  const int64_t& stop() const;
+  const int64_t& step() const;
+
+ private:
+  int64_t start_;
+  int64_t stop_;
+  int64_t step_;
+};
+
+CAFFE2_API std::ostream& operator<<(std::ostream& stream, const Slice& slice);
+
+// `at::indexing::TensorIndex` is used for converting C++ tensor indices such as
+// `{None, "...", Ellipsis, 0, true, {1, None, 2}, torch::tensor({1, 2})}`
+// into its equivalent `std::vector<TensorIndex>`, so that further tensor indexing
+// operations can be performed using the supplied indices.
+//
+// There is one-to-one correspondence between Python and C++ tensor index types:
+// Python                  | C++
+// -----------------------------------------------------
+// `None`                  | `at::indexing::None`
+// `Ellipsis`              | `at::indexing::Ellipsis`
+// `...`                   | `"..."`
+// `123`                   | `123`  
+// `True` / `False`        | `true` / `false`
+// `:`                     | `{}` / `{None, None}`
+// `::`                    | `{}` / `{None, None, None}`
+// `1:`                    | `{1, None}`
+// `1::`                   | `{1, None, None}`
+// `:3`                    | `{None, 3}`
+// `:3:`                   | `{None, 3, None}`
+// `::2`                   | `{None, None, 2}`
+// `1:3`                   | `{1, 3}`
+// `1::2`                  | `{1, None, 2}`
+// `:3:2`                  | `{None, 3, 2}`
+// `1:3:2`                 | `{1, 3, 2}`
+// `torch.tensor([1, 2])`) | `torch::tensor({1, 2})`
+struct CAFFE2_API TensorIndex {
+  // Case 1: `at::indexing::None`
+  TensorIndex(c10::nullopt_t);
+
+  // Case 2: "..." / `at::indexing::Ellipsis`
+  TensorIndex(at::indexing::EllipsisIndexType);
+  TensorIndex(const char *str);
+
+  // Case 3: Integer value
+  TensorIndex(int64_t integer);
+  TensorIndex(int integer);
+
+  // Case 4: Boolean value
+  TensorIndex(bool boolean);
```

Comment:
It might be a good idea to make sure that you don't do implicit conversions into boolean, since this can lead to some nasty bugs; e.g., your current constructor accepts a `std::shared_ptr<Tensor>`. You can do that with a template. (cc @zdevito)